### PR TITLE
Feature media library cropper

### LINKED ISSUE #2033: Add cropper to media library
## Type

- Feature

## Problem description

Some people want to crop their pictures, would be a good feature to add

### PR DESCRIPTION
**Pull request by carakas**
## Type

- Feature

## Resolves the following issues

fixes #2033 

## Pull request description

This adds a cropper to the media library with the option for the developer to force a certain aspect ratio
screen shot 2017-06-01 at 11 00 08

## TODO

- [x] limit max with and height to 3000
- [x] make it possible to use the cropper or not when no aspect ratio is given
- [x] fix scrolling on small screens
- [x] add reset button
- [x] fix strange diff after uploading multiple images

### CODE REVIEW

File: docs/07. media library/01. configuration.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Configuration
+
+The media library uses the [Liip Imagine Bundle](http://symfony.com/doc/1.0/bundles/LiipImagineBundle/index.html) to serve the images.
+
+The media library in fork comes with 5 predefined filters. For more information about this you can check the [official documentation](http://symfony.com/doc/1.0/bundles/LiipImagineBundle/filters.html).
```
Comment (jeroendesloovere) [inline docs/07. media library/01. configuration.md]: Suggestions to make things specific:
- "For more information about **how to configure filters** you can ..."
- [official **Liip Imagine Bundle** documentation]

File: docs/07. media library/02. integrating in your module.md [hunk 2]
```diff
@@ -0,0 +1,282 @@
+# How to integrate the Fork CMS MediaLibrary Module into a custom Fork CMS Module?
+
+> The example below shows you how easy it is to add custom galleries to your custom module. We show you how to add an "images" and "trailers" MediaGroup.
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: Suggestion:
- to add **media groups** to your ...

File: docs/07. media library/02. integrating in your module.md [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+{# Only include the BackendMediaGroupsHelper once #}
+{% include "MediaLibrary/Resources/views/BackendMediaGroupsHelper.html.twig" %}
+
+{{ form_widget(form.mediaGroup, {'mediaGroup': mediaGroup}) }}
+```
+
+Example: `src/Backend/Modules/MediaGalleries/Layout/Templates/Add.html.twig` and `src/Backend/Modules/MediaGalleries/Layout/Templates/Edit.html.twig`
+
+## Integration in Frontend
+
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: \### First possibility: integrate in the PHP action class.

File: docs/07. media library/02. integrating in your module.md [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+                'Slider',
+                $this->blogArticle->getImageMediaGroup()->getId(),
+                'MyCustomOptionalTitle'
+            )
+        );
+    }
+}
+```
+> You can optionally give in a fourth parameter $module (default $module = "Media"), so you can create a Media widget in your own Custom module.
+
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: \### Second possibility: use the available TwigFunction `media_library_widget`.

 ```twig
{# Slider widget #}
{{ media_library_widget('Slider', blogArticle.mediaGroup.id) }}
```
> Possible values in the core are: 'Slider', 'Lightbox', 'Listing', 'OneImage', 'OneRandomImage'. But you can easily create your own by just duplicating one of these files...

File: docs/07. media library/02. integrating in your module.md [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+```php
+$this->tpl->assign('mediaGroup', $form->getData()->mediaGroup);
+```
+
+Example: `src/Backend/Modules/MediaGalleries/Actions/Edit.php`
+
+### In your Add and Edit template
+
+This will result in:
+![Screenshow of the Backend Meda Helper](./assets/integration-example-01.png)
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: This screenshot is outdated (no Bootstrap) + in the wrong dutch (should be in english)

File: docs/07. media library/02. integrating in your module.md (hunk 2 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: I don't get it

File: docs/07. media library/02. integrating in your module.md (hunk 5 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 0d62f9fc9

File: docs/07. media library/01. configuration.md (hunk 1 above)
Comment (carakas) [inline docs/07. media library/01. configuration.md]: fixed in 79a2b12f4

File: docs/07. media library/02. integrating in your module.md (hunk 2 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 79a2b12f4

File: docs/07. media library/02. integrating in your module.md (hunk 3 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 79a2b12f4

File: docs/07. media library/02. integrating in your module.md (hunk 4 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 79a2b12f4

File: docs/01. installation/01. installation.md [hunk 6]
```diff
@@ -9,10 +9,10 @@ PHP Developers using Composer can use this command to install the latest stable
 Alternatively, you can install Fork CMS by downloading it from the website.
 Before downloading, check if your server meets the [minimum requirements](docs/03. minimum_requirements.md) for
 Fork CMS. If it does, create a MySQL database and remember the credentials. Next,
-[download](http://www.fork-cms.com/download) the lastest release and
+[download](http://www.fork-cms.com/download) the latest release and
```
Comment (WouterSioen) [inline docs/01. installation/01. installation.md]: Nice catch!

File: docs/01. installation/01. installation.md [hunk 7]
```diff
@@ -9,10 +9,10 @@ PHP Developers using Composer can use this command to install the latest stable
...
 [unpack](http://en.wikipedia.org/wiki/Tar_%28file_format%29) the folder.
 
-![Download started](https://github.com/forkcms/documentation/raw/master/01.%20installation/assets/started_download.png)
+![Download started](./assets/started_download.png)
```
Comment (WouterSioen) [inline docs/01. installation/01. installation.md]: This doesn't really belong in this pr, but still good to make it relative.

File: docs/07. media library/02. integrating in your module.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+ *      cascade="persist",
+ *      orphanRemoval=true
+ * )
+ * @ORM\JoinColumn(
+ *      name="mediaGroupId",
+ *      referencedColumnName="id",
+ *      onDelete="cascade"
+ * )
+ */
+protected $group;
```
Comment (WouterSioen) [inline docs/07. media library/02. integrating in your module.md]: This should be protected for doctrine? Or can it be private?

File: docs/07. media library/02. integrating in your module.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+```
+> You can replace `lbl.MediaConnected` by your own backend label, f.e.: 'Screenshots', 'Trailers', 'Preview images', ...
+
+Example: `Backend/Modules/MediaGalleries/Domain/MediaGallery/MediaGalleryType.php#L92`
+
+### In your Create Command class
+
+```php
+use Backend\Modules\MediaLibrary\Domain\MediaGroup\MediaGroup;
+use Backend\Modules\MediaLibrary\Domain\MediaGroup\Type as MediaGroupType;
```
Comment (WouterSioen) [inline docs/07. media library/02. integrating in your module.md]: Is there already documentation about the command bus? Otherwise `your create command` might be a little abstract for most users.

File: docs/07. media library/02. integrating in your module.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+
+Example: `src/Backend/Modules/MediaGalleries/Domain/MediaGallery/Command/CreateMediaGallery.php`
+
+Example line: `src/Backend/Modules/MediaGalleries/Domain/MediaGallery/Command/CreateMediaGallery.php#L64`
+
+### In your Create Command Handler class
+
+Don't forget to inject your new MediaGroup into your CustomEntity.
+
+Example line: `src/Backend/Modules/MediaGalleries/Domain/MediaGallery/Command/CreateMediaGalleryHandler.php#L35`
```
Comment (WouterSioen) [inline docs/07. media library/02. integrating in your module.md]: Isn't it better to put code in here? What if somebody changes something in the referenced file?

File: docs/07. media library/02. integrating in your module.md [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+
+Possible values in the core are: 'Slider', 'Lightbox', 'Listing', 'OneImage', 'OneRandomImage'. But you can easily create your own by just duplicating one of these files.
+
+## Database changes
+
+> We have chosen to have a tight-coupled connection between your "custom entity" and the "MediaGroup", which holds all the connected media items.
+
+### When using Doctrine entities
+
+> Since Fork CMS 4, more and more modules are going to use Doctrine for saving their objects (= Called Entities) to their database.
```
Comment (WouterSioen) [inline docs/07. media library/02. integrating in your module.md]: Called entities => called should nog be capitalized

File: docs/07. media library/03. console commands.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+Will delete the `/src/Frontend/Files/Cache/media_library_*` folders. Which means that all frontend thumbnails will be generated next time on load.
+
+By default the backend thumbnails will remain cached. If you also want to remove the generated backend thumbnails you need to append the ` --all` option.
+
+
+## Delete all MediaItem entities
+
+```bash
+app/console media_library:delete:items
```
Comment (WouterSioen) [inline docs/07. media library/03. console commands.md]: Wouldn't `delete:unused` be a clearer name?

File: package.json [hunk 13]
```diff
@@ -9,6 +9,7 @@
   "devDependencies": {
     "bootstrap-sass": "~3.3",
     "bootstrap-tagsinput": "~0.7",
+    "cropper": "^3.0.0-rc.2",
```
Comment (WouterSioen) [inline package.json]: Not a stable package?

File: src/Backend/Modules/MediaLibrary/Ajax/MediaItemFindAll.php [hunk 14]
```diff
@@ -23,20 +24,31 @@ public function execute(): void
...
+                'media' => $this->loadMediaItems($mediaFolder, $aspectRatio),
                 'folder' => $mediaFolder !== null ? $mediaFolder->getId() : null,
                 'tab' => $this->selectedTab,
             ]
         );
     }
 
+    private function getAspectRatio(): ?AspectRatio
+    {
+        $aspectRatio = $this->get('request')->request->get('aspect_ratio');
```
Comment (WouterSioen) [inline src/Backend/Modules/MediaLibrary/Ajax/MediaItemFindAll.php]: In think there are methods on request to get your variable casted (as a float in this case if I'm following the logic). This would allow you to avoid the `=== 'false'` check.

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js [hunk 15]
```diff
@@ -799,6 +799,196 @@ jsBackend.mediaLibraryHelper.group =
...
+    isCropping: false,
+
+    passToCropper: function(resizeInfo, resolve, reject) {
+        jsBackend.mediaLibraryHelper.cropper.cropperQueue.push({
+            'resizeInfo': resizeInfo,
+            'resolve': resolve,
+            'reject': reject,
+        });
+
+        // If the cropper is already handling the queue we don't need to start it a second time.
```
Comment (WouterSioen) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: It's a bit nitpicking, but this looks like the only place where you start your comment with a capital.

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js [hunk 16]
```diff
@@ -799,6 +799,196 @@ jsBackend.mediaLibraryHelper.group =
...
+        if (jsBackend.mediaLibraryHelper.cropper.isCropping) {
+            return;
+        }
+
+        jsBackend.mediaLibraryHelper.cropper.isCropping = true;
+        let $dialog = jsBackend.mediaLibraryHelper.cropper.getDialog();
```
Comment (WouterSioen) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: Is let already supported in all browsers?

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js [hunk 17]
```diff
@@ -889,6 +1076,24 @@ jsBackend.mediaLibraryHelper.upload =
     },
 
     /**
+    * Configure the uploader to trigger the cropper
```
Comment (WouterSioen) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: A missing space before the `*` if you want to make it consistent with the phpdocs

File: docs/07. media library/03. console commands.md (hunk 12 above)
Comment (jeroendesloovere) [inline docs/07. media library/03. console commands.md]: No, that can't be used, because there is a "force" delete which actually can delete "all" images.

File: docs/07. media library/02. integrating in your module.md (hunk 8 above)
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: can be private

File: docs/07. media library/02. integrating in your module.md (hunk 9 above)
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: There can be refered to the commands for creating a MediaGallery...?

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 15 above)
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: Starting comments with a capital seems more logic to me.
But indeed most of the comments in this mediaLibraryHelper.js (created by me) have inconsistent first letter case...

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 16 above)
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: Never used that yet... browser support only from IE11, hmmm I suggest using var

File: docs/07. media library/02. integrating in your module.md (hunk 8 above)
Comment (WouterSioen) [inline docs/07. media library/02. integrating in your module.md]: @carakas any reason you made it protected here?

File: docs/07. media library/03. console commands.md (hunk 12 above)
Comment (WouterSioen) [inline docs/07. media library/03. console commands.md]: I find `delete:items` not really a clear name for a command deleting only unused items.

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 15 above)
Comment (WouterSioen) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: There might be a standard with rules/preferences about this.

File: docs/07. media library/02. integrating in your module.md (hunk 8 above)
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: He just copy/pasted my documentation, there it was protected...

File: docs/07. media library/03. console commands.md (hunk 12 above)
Comment (jeroendesloovere) [inline docs/07. media library/03. console commands.md]: Like I said, the method indeed deletes "unused items", but if you throw an extra argument at it, it deletes "all items". So renaming to delete:unused is wrong.

Comment (jeroendesloovere) [inline docs/07. media library/03. console commands.md]: The "unused" items is just something to prevent accidentally deleting all...

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js [hunk 18]
```diff
@@ -841,15 +1032,9 @@ jsBackend.mediaLibraryHelper.upload =
             validation: {
                 allowedExtensions: jsBackend.data.get('MediaLibrary.mediaAllowedExtensions')
             },
-            scaling: {
-                includeExif: true,
-                sendOriginal: false,
-                sizes: [
-                    {name: "", maxSize: 3000}
```
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: We need a limit to the image size, otherwise "thumbnail generating" will reach maximum memory limit and things will break.
=> In your code you have written `{name: "", maxSize: 1} // used to trigger the cropper`

I had set it to 3000 (= pixels) after some testing rounds.

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js [hunk 19]
```diff
@@ -889,6 +1076,24 @@ jsBackend.mediaLibraryHelper.upload =
...
 
     /**
+    * Configure the uploader to trigger the cropper
+    */
+    getScalingConfig: function() {
+        return {
+            includeExif: false, // needs to be false to prevent issues during the cropping process, it also is good for privacy reasons
+            sendOriginal: false,
+            sizes: [
+                {name: "", maxSize: 1} // used to trigger the cropper
```
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: Please make comment more clear.

File: src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml [hunk 20]
```diff
@@ -266,6 +266,22 @@
         <translation language="en"><![CDATA[back to overview]]></translation>
         <translation language="nl"><![CDATA[terug naar overzicht]]></translation>
       </item>
+      <item type="label" name="Crop">
```
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml]: This translation belongs to < Backend >< Core >, otherwise the label isn't visible in other modules. Example of this in screenshot:
schermafbeelding 2017-06-07 om 16 40 48

File: src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml [hunk 21]
```diff
@@ -266,6 +266,22 @@
...
+        <translation language="en"><![CDATA[crop]]></translation>
+        <translation language="nl"><![CDATA[bijsnijden]]></translation>
+      </item>
+      <item type="label" name="RotateLeft">
```
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml]: This translation belongs to < Backend >< Core >, otherwise the label isn't visible in other modules.

File: src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml [hunk 22]
```diff
@@ -266,6 +266,22 @@
...
+        <translation language="en"><![CDATA[rotate left]]></translation>
+        <translation language="nl"><![CDATA[draai naar links]]></translation>
+      </item>
+      <item type="label" name="RotateRight">
```
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml]: This translation belongs to < Backend >< Core >, otherwise the label isn't visible in other modules.

File: src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml [hunk 23]
```diff
@@ -266,6 +266,22 @@
...
+        <translation language="en"><![CDATA[rotate right]]></translation>
+        <translation language="nl"><![CDATA[draai naar rechts]]></translation>
+      </item>
+      <item type="label" name="CropImage">
```
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Installer/Data/locale.xml]: This translation belongs to < Backend >< Core >, otherwise the label isn't visible in other modules.

**Comment by jeroendesloovere** [comment]
There is a bug in /src/Backend/Core/Engine/DataGridFunctions.php in the `showImage` method.
Because of this bug the "backend thumbnails" aren't shown in the MediaItemIndex.

This is the function that works:
```
    public static function showImage(
        string $path,
        string $image,
        string $title = '',
        string $url = null,
        int $width = 0,  // <-- change here
        int $height = 0,  // <-- change here
        string $filter = null
    ): string {
        $imagePath = $path . '/' . $image;

        if ($filter !== null) {
            $imagePath = BackendModel::get('liip_imagine.cache.manager')->getBrowserPath($imagePath, $filter);
        }

        $html = '<img src="' . $imagePath . '" alt="' . $title . '"';

        if ($width !== 0) {  // <-- change here
            $html .= ' width="' . $width . '"';
        }

        if ($height !== 0) {  // <-- change here
            $html .= ' height="' . $height . '"';
        }

        $html .= ' />';

        if ($url !== null) {
            $html = '<a href="' . $url . '" title="' . $title . '">' . $html . '</a>';
        }

        return $html;
    }
```

File: package.json (hunk 13 above)
Comment (carakas) [inline package.json]: still waiting for it, will update when it arrived

File: src/Backend/Modules/MediaLibrary/Ajax/MediaItemFindAll.php (hunk 14 above)
Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Ajax/MediaItemFindAll.php]: nah, I checked, float is missing from the list

File: docs/07. media library/02. integrating in your module.md (hunk 8 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: 16fe0b48c

File: docs/07. media library/02. integrating in your module.md (hunk 11 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: 16fe0b48c

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 17 above)
Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: fixed in c0cc5cd4d

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 16 above)
Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: fixed in f4d421500

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 17 above)
Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: fixed in c0cc5cd4d

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 16 above)
Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: fixed in f4d421500

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 18 above)
Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: it needs to be 1 to trigger the cropper

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 19 above)
Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: fixed in f2f4862dd

File: src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js (hunk 18 above)
Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: Why exactly?
1 will give lots of user errors, since I tested this for a while and 3000 was a nice compromise between, working images and failing images...

Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: I updated the comment, I can maybe see if I can add something to manually check on that but the way you say to fix it will break the cropper, that's why the comment is there

Comment (jeroendesloovere) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: @carakas Can you tell me why it breaks?

Comment (carakas) [inline src/Backend/Modules/MediaLibrary/Js/MediaLibraryHelper.js]: I already did

File: docs/07. media library/02. integrating in your module.md (hunk 9 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: @WouterSioen there is no documentation for the commandbus afaik,after the release of fork 5 maybe it would be a good idea to focus on getting the documentation a bit more up to date

File: docs/07. media library/02. integrating in your module.md (hunk 10 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 25cf448c1

**Comment by jeroendesloovere** [comment]
@carakas Hi, just checking to know if you changed the following:
By default the cropper should be disabled.

**Comment by carakas** [comment]
Disabling it by default is your opinion, I'm wondering what the others think about that

**Comment by jeroendesloovere** [comment]
@carakas I created this module like 6 years ago. Never needed any cropping.
The frontend can crop images as well using LiipImagineBundle filter sets.
If you upload 30 images, it's very annoying, ...
Reasons enough to disable this cropper, because it's a "small use case scenario".

**Comment by jeroendesloovere** [comment]
@carakas Correct me if wrong, but there is an option to "disable the cropper"?

**Comment by carakas** [comment]
I still have some changes that have been scheduled so to better reflect this I changed the title to [WIP]

**Comment by carakas** [comment]
Should be finished now

File: docs/07. media library/02. integrating in your module.md [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+```php
+use Backend\Modules\MediaLibrary\Domain\MediaGroup\MediaGroupType;
+```
+
+```php
+$builder->add(
+    'mediaGroup',
+    MediaGroupType::class,
+    [
+        'label' => 'lbl.MediaConnected',
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: Add:
`'constraints' => [new Valid()],`

File: docs/07. media library/02. integrating in your module.md [hunk 25]
```diff
@@ -0,0 +1,294 @@
...
+ *      cascade="persist",
+ *      orphanRemoval=true
+ * )
+ * @ORM\JoinColumn(
+ *      name="mediaGroupId",
+ *      referencedColumnName="id",
+ *      onDelete="cascade"
+ * )
+ */
+private $group;
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: Rename to a more logical `$mediaGroup:`

File: docs/07. media library/02. integrating in your module.md [hunk 26]
```diff
@@ -0,0 +1,294 @@
...
+```
+
+```php
+/**
+ * Gets the value of group.
+ *
+ * @return MediaGroup
+ */
+public function getGroup()
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: Change `getGroup()` to `getMediaGroup(): MediaGroup`

File: docs/07. media library/02. integrating in your module.md [hunk 27]
```diff
@@ -0,0 +1,294 @@
...
+ *      name="mediaGroupId",
+ *      referencedColumnName="id",
+ *      onDelete="cascade"
+ * )
+ */
+private $group;
+```
+
+```php
+/**
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: Remove this PHPDoc (since 7.1 is self-explaining)

File: docs/07. media library/02. integrating in your module.md [hunk 28]
```diff
@@ -0,0 +1,294 @@
...
+ * Gets the value of group.
+ *
+ * @return MediaGroup
+ */
+public function getGroup()
+{
+    return $this->group;
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: Change to
`return $this->mediaGroup;`

File: docs/07. media library/02. integrating in your module.md [hunk 29]
```diff
@@ -0,0 +1,294 @@
...
+```php
+/**
+ * @var MediaGroup
+ */
+public $mediaGroup;
+```
+
+**In your __construct()**
+```php
+$this->mediaGroup = $customEntity->getGroup();
```
Comment (jeroendesloovere) [inline docs/07. media library/02. integrating in your module.md]: Change to 
`$this->mediaGroup = $customEntity->getMediaGroup();`

Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 6da2e49c2

File: docs/07. media library/02. integrating in your module.md (hunk 24 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 6da2e49c2

File: docs/07. media library/02. integrating in your module.md (hunk 28 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 6da2e49c2

File: docs/07. media library/02. integrating in your module.md (hunk 26 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 6da2e49c2

File: docs/07. media library/02. integrating in your module.md (hunk 27 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 6da2e49c2

File: docs/07. media library/02. integrating in your module.md (hunk 25 above)
Comment (carakas) [inline docs/07. media library/02. integrating in your module.md]: fixed in 6da2e49c2
